Ensure all packages declare `package-info.java` with null-safety annotations

edyda99 opened the pull request:
## Overview

This commit adds the `JavadocPackage` Checkstyle module to the project's configuration file (`checkstyle.xml`). The module is configured to emit a warning message when a `package-info.java` file is missing, rather than throwing an exception. In addition, suppressions have been added to prevent the `JavadocPackage` checks from being applied to packages outside of the `src/main` directory. These suppressions ensure that the module only checks the packages that are relevant to the project and prevents false positives from being reported.

## Deliverables

- [x] Configure the [`JavadocPackage`](https://checkstyle.sourceforge.io/config_javadoc.html#JavadocPackage) Checkstyle module to require `package-info.java` files for all packages under `src/main`.
- [x] Determine if it is possible to require that `package-info.java` files include null-safety annotations – for example, via Checkstyle.
- [x] If it's possible to require that `package-info.java` files include null-safety annotations, configure the necessary infrastructure.
- [x] Introduce missing `package-info.java` files with package-level Javadoc and null-safety annotations.

simonbasle reviewed (commented):
thanks for the suggestion. Can you provide more details about the choices you made for checkstyle modules?

simonbasle commented:
(didn't see the issue reference, which probably provides some context. my bad)

simonbasle commented:
@edyda99 I think the `RegexpJavaSingleLine` variant does the job (unless you've spotted occurrences of multilines nullability annotations?) The other one is also harder to read and has its own caveats with comments, so I'd go with the first one.

edyda99 commented:
Hi @simonbasle, I will go with RegexpJavaSingleLine.
No, I didn't spot any occurrences of multilines nullability, but it was about preventing it from happening.
So should I remove the whitespace warning? ( I would suggest keeping it and increasing the severity to "**error**", so whenever someone writes `@\n<annotation_name>` an error will be thrown)

simonbasle commented:
> ( I would suggest keeping it and increasing the severity to "**error**", so whenever someone writes `@\n<annotation_name>` an error will be thrown)

Perhaps a simpler one line regexp like `@\s*$` (at then any number of whitespace then the end of the line) would fit the bill? That said, I think this PR can focus on the actual null-safety annotations.

simonbasle commented:
>     * Do you suggest doing micro commits for each missing package-info.java?

Not necessarily, if that's the only thing in the commit.

>     * What do I need to rely on before adding null-safety annotations on a newly added package-info?

It depends on how many package-info are missing. I would locally prepare the change and run some static analysis trying to find null-safety issues raising from that change. If the build doesn't break, I'm not sure I would make it a goal of this PR to fix the null-safety warnings though. Wdyt @sbrannen (also cc @sdeleuze)

sbrannen commented:
> > * Do you suggest doing micro commits for each missing package-info.java?
> 
> Not necessarily, if that's the only thing in the commit.

I agree with @simonbasle.

Ideally the final result will be 3-4 commits.

- enforce that `package-info.java` files exist and contain Javadoc
  - build should fail
- introduce missing `package-info.java` files
  - build should pass
- enforce that `package-info.java` files declare `@NonNullApi` and `@NonNullFields`
  - build should fail
- introduce missing `@NonNullApi` and `@NonNullFields` declarations
  - build should pass

Of course, we may squash all commits in the PR into a single commit if that turns out to be easier.

> > * What do I need to rely on before adding null-safety annotations on a newly added package-info?

I don't understand the question. Can you please expound?

> It depends on how many package-info are missing. I would locally prepare the change and run some static analysis trying to find null-safety issues raising from that change. If the build doesn't break, I'm not sure I would make it a goal of this PR to fix the null-safety warnings though. Wdyt @sbrannen (also cc @sdeleuze)

Ensuring that proper null-safety semantics are enforced _within_ code in newly annotated packages is beyond the scope of this PR/issue. I'll create a follow-up issue to address that.

sbrannen commented on src/checkstyle/checkstyle.xml:
```diff
@@ -239,5 +239,21 @@
+			<property name="severity" value="warning"/>
+			<property name="minimum" value="1"/>
+			<property  name="maximum" value="5"/>
+			<property name="message" value="package-info.java is missing null-safety annotations  or has null-safe annotations made up of multiple lines. Please check."/>
```
```suggestion
			<property name="message" value="package-info.java is missing required null-safety annotations @NonNullApi and @NonNullFields."/>
```

Spring's `@NonNullApi` and `@NonNullFields` annotations should never span multiple lines.

sbrannen reviewed (changes requested):
Thanks for working on this!

It's coming along nicely.

I've requested a few minor changes.

After that, all that's left is ensuring that the build fails for violations and that all missing `package-info.java` files have been created (so that the build passes again).

sbrannen commented on src/checkstyle/checkstyle.xml:
```diff
@@ -239,5 +239,21 @@
+
+	<!--package-info checker -->
+	<module name="JavadocPackage">
+		<property name="severity" value="warning"/>
```
We also want these violations to fail the build.

sbrannen commented:
You also need to suppress violations in the `framework-docs` module, since that source code is only used in the reference manual.

edyda99 committed:
Enforce package-info.java files on /src/main/ and framework-docs module and specific packages inside spring-core module

This commit updates the project's checkstyle configuration to require the existence of a package-info.java file for all packages within the src/main directory while excluding framework-docs module and certain packages inside spring-core module. A missing package-info.java will result in emitting an exception.

edyda99 committed:
Enforce package-info.java files on /src/main/ and framework-docs module and specific packages inside spring-core module

This commit updates the project's checkstyle configuration to require the existence of a package-info.java file for all packages within the src/main directory while excluding framework-docs module and certain packages inside spring-core module. A missing package-info.java will result in emitting a warning.

edyda99 committed:
Enforce package-info.java files on /src/main/ and framework-docs module and specific packages inside spring-core module

This commit updates the project's checkstyle configuration to require the existence of a package-info.java file for all packages within the src/main directory while excluding framework-docs module and certain packages inside spring-core module. A missing package-info.java will result in emitting a warning. Additionally, two RegexpSinglelineJava modules have been added to check that package-info.java files contain null-safety annotations @NonNullApi and NonNullFields and throws exception in case one of them or both are missing.

edyda99 committed:
Enforce package-info.java files on /src/main/ and framework-docs module and specific packages inside spring-core module

This commit updates the project's checkstyle configuration to require the existence of a package-info.java file for all packages within the src/main directory while excluding framework-docs module and certain packages inside spring-core module. A missing package-info.java will result in emitting a warning. Additionally, two RegexpSinglelineJava modules have been added to check that package-info.java files contain null-safety annotations @NonNullApi and NonNullFields.

edyda99 commented:
Hi again,

First sorry for the continuous Force pushes, but I wanted to reduce my commits.
I reduced them to 4 micro commits, with the last one containing the full commit message.
All comments are resolved.

The merge request is ready for the final review

Thank you again!

edyda99 commented:
@sbrannen Ok Great
Could you suggest something else for me?

sbrannen commented:
This has been merged into `main` in 268e3fec9921653de17ab2f495039239f6b6152e and 7e32f504b021015ca8a7eaed7280453fad724b0f and revised and completed in 99e54fec3ac8dfedd2537fde5f19fbcc4f1f6202.

The latter turned out to be a bit more work than expected. 👀 

Thanks again for the contribution, @edyda99!
